clarify delegation proof requires `capabilityChain` and update examples 3 & 4 to have a `capabilityChain` (#78)
w3c-ccg/zcap-spec: decision/20260807-clarify-proof-capabilityChain.md

# Clarify when `proof.capabilityChain` is required, and ensure examples use it when required

Authors
* bengo.is

## Introduction

[example 1][] and [example 7][] are capability delegations that have a `proof.capabilityChain` property.

Before this decision
* the property `capabilityChain` wasn't described anywhere in the spec, just used in examples 1 and 7 (but not 3 and 4)
* no examples offered an example of what a `proof.capabilityChain` value looks like for a zcap that is a delegation of a delegation

After this decision
* the property name `capabilityChain` is explicitly used in the relevant capability delegation proof requirement. This avoid the ambiguity by clearly requiring use of `capabilityChain`, offering clear guidance for examples
* update examples 3 and 4 to use `proof.capabilityChain` as required

> * examples 3 and 4 show what a `proof.capabilityChain` value looks like for a zcap that is a delegation of a delegation: the 

Reviewer: The sentence at the end was cut short.

> * examples 3 and 4 show what a `proof.capabilityChain` value looks like for a zcap that is a delegation of a delegation: the 

Author: ```suggestion
* examples 3 and 4 show what a `proof.capabilityChain` value looks like for a zcap that is a delegation of a delegation. 
```

In thinking through what to do about it, I am left with some questions, and a thought that maybe the spec's examples are inconsistent with its recommendations, and the result is a lack of clarity about how delegations and invocations should be proven.

The examples should be consistent and conformant in how they use `proof.capabilityChain`. Before this decision, they were not, 

## Background: Spec Text

* In [Delegation through Capability Chains](https://w3c-ccg.github.io/zcap-spec/#delegation)

>   * > A capability document that is not the target MUST also have a parentCapability property which either points at the target or another capability document. A series of capability chained together in this way is called a "capability chain" and is how delegation of capabilities are handled in zcaps.

Reviewer: ```suggestion
  * > A capability document that is not the target MUST also have a parentCapability property which either points at the target or another capability document. A series of capabilities chained together in this way is called a "capability chain" and is how delegation of capabilities is handled in zcaps.
```

Author: Thank you!

To
> A delegated zcap can only be invoked by submitting the entire zcap.
> 
> A delegated zcap MUST have a capability delegation proof.
> A capability delegation proof MUST have a `capabilityChain` property whose value is a capability ancestors array.
> A capability ancestors array MUST be an array that includes the root zcap using its ID (i.e., by reference only, not embedded) and every other delegated zcap in its ancestry must be referenced by ID except for the parent delegated zcap, which MUST be fully embedded.

This change clarifies that the capability delegation proof should use a `capabilityChain` property to include the array.

The change also uses "capability ancestors array" instead of "capability delegation chain" to avoid the current challenge of confusing "capability delegation chain" (array in proof) with "capability chain". The new term was chosen because
* it represent that this array does not have the full chain, only the chain of ancestors capabilities of the proven delegated capability

> * it clarifies the data type is an array, to distinguish it from the other, currently more prevalent in the spec, "capability chain" that is not an array but a linked list.

Reviewer: We may want to change our use of “capability chain” in the spec if it does not reflect the representation captured in `capabilityChain`. If “capability chain” and `capabilityChain` don’t refer to the same thing, this might cause confusion for new readers. I don’t mind “capability ancestors array” at face value (though I would prefer “capability ancestor array” without the “s”), but I am concerned about the misalignment (albeit slight) around the semantics.

One solution would be to rename `capabilityChain`, but that is normative and instantly breaks all implementations. Instead, my suggestion would be to keep our use of “capability chain” (instead of “capability ancestors array”), but use it to describe `capabilityChain` (instead of whatever notion it is describing now), for the sake of consistency. Then we would just have to choose a different name to describe whatever “capability chain” is currently describing.

Open to other thoughts though…

Author: I am sympathetic to the concern. I'm open to bigger changes in the future, e.g. renaming `capabilityChain` when we improve/formalize the normative text on proof generation / verification and hopefully rationalize it against the data integrity specs that we can no normatively reference, unlike when a lot of the current text was written.
I think eventually I or you will be able to make a PR to improve the design and normative intent, but for now this PR only meant to make non-normative change in the hopes to reduce ambiguity of the spec.

Let's discuss the broader design space on the list or meeting time.

This change adjusts normative text for clarity, but it is not intended as a normative change.
The existing normative text is too ambiguous for the requirement to be satisfied.

## Problem: Example 1 `parentCapability` aims to be root capability, but uses a URI that implies a delegated capability

[Example 1][] (many properties omitted) is like
```json
{
  "id": "https://whatacar.example/a-fancy-car/proc/7a397d7b",
  // Since this is the first delegated capability, the parentCapability
  // points to the target this capability will operate against
  // (in this case, Alyssa's Car)
  "parentCapability": "https://whatacar.example/a-fancy-car",
  "proof": {
    "capabilityChain": [
      "https://whatacar.example/a-fancy-car"
    ]
  }
}
```

> The comment above example 1 `parentCapability` implies that the identified `parentCapability` is meant to be a root capability, and it is identified withan `https:` URL.

Reviewer: ```suggestion
The comment above example 1 `parentCapability` implies that the identified `parentCapability` is meant to be a root capability, and it is identified with an `https:` URL.
```

The `proof.capabilityChain` includes this same URL as the first item, also hinting that the intention is for the parentCapability to be a root capability zcap.

However, later the [Root Capability](https://w3c-ccg.github.io/zcap-spec/#root-capability) section requires
> A root zcap MUST have an id that is a string that expresses a URN

> Because the currently parentCapability URL does not satisfy requirements of a root capability `id`, the delegation's `proof.capabilityChain` does not appear to satisfy the requirement

Reviewer: ```suggestion
Because the parentCapability URL currently does not satisfy requirements of a root capability `id`, the delegation's `proof.capabilityChain` does not appear to satisfy the requirement
```

This would be satisfied by the other proposals to add `proof.capabilityChain` to Examples 3 and 4.

## Problem: Example 3 does not satisfy requirement "A delegated zcap MUST have a capability delegation proof which MUST contain the delegation chain."

[Example 3][] is a capability delegation whose `parentCapability` is `"https://whatacar.example/a-fancy-car/proc/7a397d7b"`, which is the identifier of the capability delegation from [Example 1][].

The `proof` value for the capability delegation from [Example 3][] does *not* contain a `capabilityChain` property, so it does not appear to satisfy the requirement
> A delegated zcap MUST have a capability delegation proof which MUST contain the delegation chain.

This appears to be an oversight. If so, we can add the required `capabilityChain`.

### Decision: Add `proof.capabilityChain` to Example 3

The full new Example 3 would be as follows, which has the following changes from the original example 3:

> * `proof.capabilityChain` inserted whose value is a capability capability ancestors array.

Reviewer: ```suggestion
* `proof.capabilityChain` inserted whose value is a capability ancestors array.
```

```json
{"@context": ["https://w3id.org/zcap/v1",
              "https://autopower.example/"],
 "id": "https://social.example/alyssa/caps#79795d78",

 // Pointing up the chain at the capability from which Alyssa was
 // initially gained authority
 "parentCapability": "https://whatacar.example/a-fancy-car/proc/7a397d7b",

 // Alyssa grants authority specifically to one of Ben's
 // cryptographic keys
 "controller": "https://chatty.example/ben/#key-33",

 // Alyssa adds a caveat: Ben can drive her car, unless she flips
 // the bit at this url
 "caveat": [
   {"type": "ValidWhileTrue",
    "uri": "https://social.example/alyssa/ben-can-still-drive"}],

 // Finally Alyssa signs this object with the key she was granted
 // authority with
 "proof": {
    "type": "RsaSignature2016",
    "proofPurpose": "capabilityDelegation",
    "created": "2017-03-28T06:01:25Z",
    "creator": "https://social.example/alyssa/#key-for-car",

>     "signatureValue": "...",

Reviewer: The examples seem to be mixing newer fields (e.g., `proofPurpose`) with older ones (e.g., `signatureValue` from the RSA Signature Suite 2016, which I would expect to now be `proofValue`). I know there is an intention (and maybe an issuer/PR) to address this, but I wanted to call it out here, since this PR touches on so many of the examples.

Author: Yep, good call. Let's track in https://github.com/w3c-ccg/zcap-spec/issues/85

          "proofPurpose": "capabilityDelegation",
          "verificationMethod": "https://example.com/i/alice/keys/1"
        }
      }
    ]
  }
}
```

## Problem: Example 4 does not satisfy requirement "A delegated zcap MUST have a capability delegation proof which MUST contain the delegation chain."

The `proof` value for the capability delegation from [Example 4][] does *not* contain a `capabilityChain` property, so it does not appear to satisfy the requirement
> A delegated zcap MUST have a capability delegation proof which MUST contain the delegation chain.

### Decision: Add `proof.capabilityChain` to Example 4

Make the following changes to Example 4
* add `proof.capabilityChain` array
  * the first entry is the URN of the root zcap
  * the second entry is the identifier of the delegation of the root zcap, which is `https://whatacar.example/a-fancy-car/proc/7a397d7b`
  * the third entry is the full expression of the delegation of the capability identified by the second entry. This is also the referent of the outer delegation's `parentCapability` value `"https://social.example/alyssa/caps#79795d78"`.

> This makes the example much longer, because the `capabilityChain` includes a full expression of the parent capability delegation `https://social.example/alyssa/caps#79795d78`, which itself has a `proof.capabilityChain` that includes a full expression of `https://whatacar.example/a-fancy-car/proc/7a397d7b`, which itself has a `proof.capabilityChain`. This seems to be the unavoidable if we hope the examples to be a complete demonstration of the normative requirements.

Reviewer: ```suggestion
This makes the example much longer, because the `capabilityChain` includes a full expression of the parent capability delegation `https://social.example/alyssa/caps#79795d78`, which itself has a `proof.capabilityChain` that includes a full expression of `https://whatacar.example/a-fancy-car/proc/7a397d7b`, which itself has a `proof.capabilityChain`. This seems to be unavoidable if we hope the examples to be a complete demonstration of the normative requirements.
```

* it clarifies the data type is an array, to distinguish it from the other, currently more prevalent in the spec, "capability chain" that is not an array but a linked list.

## Challenge: Requirements of "capability delegation chain" array lead to large delegation examples that are hard to read

After updating Examples 3 and 4 to satisfy the requirements of capability delegation proofs by including a proof `capabilityChain`, the examples are no longer very short.
On the other hand, their length is a more accurate example of a real-world zcap.

One way of addressing this would be to have the examples be incomplete examples, and abbreviate or omit some values.
By addressing the challenge this way
* we'd gain some way for the examples to be shorter and take up less space.
  * We'd also need to come up with a convention for these kinds of 'incomplete examples' and how to markup any abbreviations/omissions.
* we'd lose the examples being a complete example zcap.
  * this could be mitigated by including the full example zcap in an appendix, but this would mean editors are maintaining two versions of every example in different sections.
    I'd prefer to avoid that.

> Another way of addressing this would be to keep long complete examples in the text, but configure our respec tooling to use a respec plugin

Author: proof of concept https://gobengo.github.io/respec-collapsible-json/#usage
